### ISSUE TITLE
Tpetra BCRS: Detach public interface from layout of little blocks

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#179]
@trilinos/tpetra @trilinos/ifpack2 @crtrott @kyungjoo-kim @amklinv 

The public interface of Tpetra::Experimental::BlockCrsMatrix currently assumes that little blocks are row major (LayoutRight, in Kokkos terms), regardless of the execution space.  It would make sense to relax this restriction.  This would have the following advantages:
1. It would let the layout be a function of the execution space, hopefully for best performance
2. It would allow strided layouts, so that the (i,j) entries of consecutive blocks could be stored contiguously, thus making outer loop vectorization (for Intel Xeon Phi) easier

**Comment by mhoemmen** [comment]
This depends on #177.

**Comment by mhoemmen** [comment]
One critical example where we can retain backwards compatibility without tying the implementation to use a particular layout: replaceLocalValues and sumIntoLocalValues.  These methods assume row-major (LayoutRight) _input_.  However, they assume nothing about the implementation's layout of blocks!  If we retain the existing input layout, customers' code won't have to change.  This is really, really important, especially for codes that made the (admittedly suboptimal) choice of only summing in one block at a time (where it would be weird for the input to have a stride suitable for outer loop vectorization).

If we later want to relax the layout assumptions on input, we can add methods templated on the input View type.  This would have the additional advantage of preserving input memory traits.

**Comment by mhoemmen** [comment]
Problem: `BlockCrsMatrix::getLocalRowView` (that returns "views" of the row's block data as raw pointers).  No way to hide the layout there.  Users _want_ methods like this, that return a view of the row's data.  However, they do constrain the storage layout.

I would say, we provide a `Kokkos::View<impl_scalar_type***, ...>` version, and warn users that the raw pointer version uses the same layout as that version returns.  The raw pointer version still has value for users who don't want to commit to a Kokkos-ized code.

**Comment by crtrott** [comment]
Note I still want to do some performance evaluations. There is a reason that the KokkosSparse::CrsMatrix row view function does not return a Kokkos view but a different object which behaves kind of like an unmanaged view. The problem is that rows are often very short, something Kokkos Views are not optimized for. That said in a block matrix the number of scalar values per row is orders of magnitude larger, so the unmanaged View route might be just fine.

**Comment by mhoemmen** [comment]
> Note I still want to do some performance evaluations. There is a reason that the KokkosSparse::CrsMatrix row view function does not return a Kokkos view but a different object which behaves kind of like an unmanaged view. The problem is that rows are often very short, something Kokkos Views are not optimized for.

We could revive LittleBlock etc. with a different interface, if this turns out to matter.  I still would like to retain the raw pointer interface.  The nice thing about replace and sumInto is that they are agnostic of the internal storage format.

**Comment by mhoemmen** [comment]
I'm moving this to FY17.  We don't need to accomplish this in FY16.  It's a good idea, but users aren't clamoring for it.

**Comment by mhoemmen** [comment]
This is not an Ifpack2 issue.  Ifpack2 can and should access BCRS blocks through the Kokkos::View interface.

**Comment by mhoemmen** [comment]
This is a duplicate of #619, so I'll close this.
